pkg/apis/resource/validation/validation.go

```diff
@@ -893,3 +915,122 @@ func validateNetworkDeviceData(networkDeviceData *resource.NetworkDeviceData, fl
+	allErrs = append(allErrs, validateDeviceTaintRuleSpec(&deviceTaint.Spec, nil, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateDeviceTaint tests if a DeviceTaintRule update is valid.
+func ValidateDeviceTaintUpdate(deviceTaint, oldDeviceTaint *resource.DeviceTaintRule) field.ErrorList {
+	allErrs := corevalidation.ValidateObjectMetaUpdate(&deviceTaint.ObjectMeta, &oldDeviceTaint.ObjectMeta, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateDeviceTaintRuleSpec(&deviceTaint.Spec, &oldDeviceTaint.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+func validateDeviceTaintRuleSpec(spec, oldSpec *resource.DeviceTaintRuleSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	var oldFilter *resource.DeviceTaintSelector
+	if oldSpec != nil {
+		oldFilter = oldSpec.DeviceSelector // +k8s:verify-mutation:reason=clone
+	}
+	allErrs = append(allErrs, validateDeviceTaintSelector(spec.DeviceSelector, oldFilter, fldPath.Child("deviceSelector"))...)
+	allErrs = append(allErrs, validateDeviceTaint(spec.Taint, fldPath.Child("taint"))...)
+	return allErrs
+}
+
+func validateDeviceTaintSelector(filter, oldFilter *resource.DeviceTaintSelector, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if filter == nil {
+		return allErrs
+	}
+	if filter.DeviceClassName != nil {
+		allErrs = append(allErrs, validateDeviceClassName(*filter.DeviceClassName, fldPath.Child("deviceClassName"))...)
+	}
+	if filter.Driver != nil {
+		allErrs = append(allErrs, validateDriverName(*filter.Driver, fldPath.Child("driver"))...)
+	}
+	if filter.Pool != nil {
+		allErrs = append(allErrs, validatePoolName(*filter.Pool, fldPath.Child("pool"))...)
+	}
+	if filter.Device != nil {
+		allErrs = append(allErrs, validateDeviceName(*filter.Device, fldPath.Child("device"))...)
+	}
+
+	// If the selectors are exactly as before, we treat the CEL expressions as "stored".
+	// Any change, including merely reordering selectors, triggers validation as new
+	// expressions.
+	stored := false
+	if oldFilter != nil {
+		stored = apiequality.Semantic.DeepEqual(filter.Selectors, oldFilter.Selectors)
+	}
+	allErrs = append(allErrs, validateSlice(filter.Selectors, resource.DeviceSelectorsMaxSize,
+		func(selector resource.DeviceSelector, fldPath *field.Path) field.ErrorList {
+			return validateSelector(selector, fldPath, stored)
+		},
+		fldPath.Child("selectors"))...)
+
+	return allErrs
+}
+
+func validateDeviceTaint(taint resource.DeviceTaint, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, metav1validation.ValidateLabelName(taint.Key, fldPath.Child("key"))...) // Includes checking for non-empty.
+	if taint.Value != "" {
+		allErrs = append(allErrs, validateLabelValue(taint.Value, fldPath.Child("value"))...)
+	}
+	switch taint.Effect {
+	case resource.DeviceTaintEffectNoSchedule, resource.DeviceTaintEffectNoExecute:
```

Comment:
Why not the usual set pattern?

```
if taint.Effect == "" {
    field.Required(...)
} else if !validValues.Has(taint.Effect) {
    field.NotSupported(...)
} else { 
   ....
}
```

Reply:
No particular reason. Changed.

FWIW, existing code probably could be updated to follow this pattern a bit more.
I looked for `validValues` and usually found code like this:

https://github.com/kubernetes/kubernetes/blob/ed19dbc79f5bfe479b39d2685247895e735bbf70/pkg/apis/core/validation/validation.go#L3681-L3730

Only the last one uses a set, and it doesn't do the `field.Required`.

Reply:
Yeah, the reality is that it is all over the map.  If anyone doesn't see the value of declarative validation, they need to read more of these PRs. :)